### PR TITLE
Named Type Arguments & Partial Type Argument Inference

### LINKED ISSUE #10571: Allow skipping some generics when calling a function with multiple generics
Right now in TypeScript it's all or nothing when calling generic methods. You can either skip typing all the generics altogether and they will be inferred from the context (if possible), or you have to manually (re)define all of them. But the reality isn't black and white, there are also shades of gray, where we can infer types for some of the generic parameters, but not others. Currently those have to be unnecessarily verbose by forcing the programmer to explicitly re-state them.

Take a look at these 3 cases:

Case 1 - everything can be inferred - no need to call the method with `<>` definition:
```ts
function case1<A, B, C>(a: A, b: B, c: C): A {}

example(1, '2', true);
```

Compiler knows that:
 A is number
 B is string
 C is boolean

---

Case 2 - nothing can be inferred, so we need to state what A, B and C should be, otherwise they'll default to `{}`:

```ts
function case2<A, B, C>(): A {}

example<number, string, boolean>();
```

---
Case 3 - the one that's interesting to this feature request - some can be inferred, some can't:

```ts
function case3<A, B, C>(b: string, c: boolean): A {}

// incorrect type of A - left unspecified:
example('thing'); 

// correct, but unnecessarily verbose - we already know that 'thing' is a string and true is a bool
example<number, string, boolean>('thing', true);
```

Now, typing `string` in the above example isn't a big deal, but with complex scenarios, say with a method using 5 generics, where you can infer 4 of them, retyping them all seems

It would be great if we could has some way to skip re-typing the types that can automatically be inferred. Something like a special `auto` or `inferred` type, so we could write:

```ts
example<number, auto, auto>('thing', bool);
```

Or maybe even, if we want to skip the rest of the generics, only specify those up-to that point:

```ts
example<number>('thing', bool);
```

Having such a feature would solve newcomers encountering problems such as this one: http://stackoverflow.com/questions/38687965/typescript-generics-argument-type-inference/38688143

### LINKED ISSUE #20122: Granular inference for generic function type arguments
Consider the following function, based on [this StackOverflow question](https://stackoverflow.com/questions/47359713/filter-on-keyof-type-parameter-in-typescript/47359891):

```ts
// Take a type and some properties to pluck from it. Ensure only functions can be plucked.
type PluckFuncs =
  <TComponent extends {[P in TKey]: Function }, TKey extends string>
  (props: TKey[]) => any
```

Currently it is not possible to use this function ergonomically. You must always supply both type arguments, even though `TKey` can comfortably be inferred from `props`.

It should be possible to invoke a generic function and specify a named subset of the type parameters, i.e. `pluckFuncs<TComponent = { a: () => string }>(['a'])`, and have inference decide (or fail to decide) the remaining type parameters (in this case `TKey`).

In terms of the desired semantics, the inference should treat explicit type arguments no differently from the types of the function's value arguments, except for the fact that no values will be forthcoming at the invocation site. 

In other words, doing this:

```ts
type PluckFuncs =
  <TComponent extends {[P in TKey]: Function }, TKey extends string>
  (props: TKey[]) => any

declare const pluckFuncs: PluckFuncs

pluckFuncs<{ a: () => string }>(['a']) // Won't work, forced to explicitly specify `TKey`
```

should behave no differently from when I just add a dummy value argument to serve as an inference site:

```ts
type PluckFuncs =
  <TComponent extends {[P in TKey]: Function }, TKey extends string>
  (props: TKey[], dummy?: TComponent) => any

declare const pluckFuncs: PluckFuncs

pluckFuncs(['a'], undefined as { a: () => string }) // Compiler is happy with this though
```

## Aside:

The workaround one might come up with is to give `TKey` a default of `keyof TComponent`:

```ts
type PluckFuncs =
  <TComponent extends {[P in TKey]: Function }, TKey extends string = keyof TComponent>
  (props: TKey[]) => any
```

This *seems* like a sensible thing to do, but isn't really the semantics we're looking for. When you try to use it:

```ts
declare const pf: PluckFuncs

interface TestProps {a: number, b: Function, c: () => number}

// Good - missing property, should be an error
pf<TestProps>(['d'])

// Good - non-function property plucked, should be an error
pf<TestProps>(['a'])

// Bad - all plucked properties extend `Function`, but this is nevertheless an error
pf<TestProps>(['b', 'c'])

// We end up needing `TestProps` to not have any non-function properties whatsoever

// ...or to specify the second type argument explicitly
pf<TestProps, 'b' | 'c'>(['b', 'c'])
```

What's happened is that we only have one degree of freedom here. Defaulting `TKey` to a type derived from `TComponent` has ended up imposing an additional constraint on `TComponent` instead of constraining the parameter that uses `TKey` with respect to `TComponent`.

I think the behavior seen above is valid, and that default type arguments are a totally different ballgame. They shouldn't be mixed up in our problem with independent inference of type parameters when instantiating abstract function types.

### LINKED ISSUE #22631: Proposal: allow type arguments of generic being named
**TypeScript Version:**  2.7.0-dev.201xxxxx

**Search Terms:** named generic type parameters / arguments

**Code**

```ts
interface UserInfo { }
var usersMap: Map<sessionId: string, info: UserInfo>;
```

**Expected behavior:**
No error, `Map<sessionId: string, info: UserInfo>` should be equivalent to `Map<string, UserInfo>` .

**Actual behavior:**
Syntax error: Generic type 'Map<K, V>' requires 2 type argument(s).

**Playground Link:**  
https://www.typescriptlang.org/play/#src=interface%20UserInfo%20%7B%20%7D%0D%0Avar%20usersMap%3A%20Map%3CsessionId%3A%20string%2C%20info%3A%20UserInfo%3E%3B

**Rationale:**
Without a name, it is not often immediately clear what a type argument really mean. Take `Map<string, UserInfo>` as an example, is the first argument a `sessionId`, a `userId` or something else? In contrast, ` Map<sessionId: string, info: UserInfo>` or ` Map<sessionId: string, UserInfo>` is very clear.

The name of a type argument doesn't affect the actual type, just like the name of argument of a function type.

### PR DESCRIPTION
**Pull request by weswigham**
With this PR, we allow named type arguments, of the form `Identifier = Type` anywhere a type argument is expected. This looks like so:

```ts
const instance1 = new Foo<T = number, U = string>(0, "");
const result1 = foo<T = number, U = string>(0, "");
const tagged1 = tag<T = number, U = string>`tags ${12} ${""}`;
const jsx1 = <Component<T = number, U = string> x={12} y="" cb={props => void (props.x.toFixed() + props.y.toUpperCase())} />;
type A = Foo<T = number, U = string>;
```

These arguments do not need to come in any particular order, but _must_ come after all positional type arguments. When you've used a named type argument, you may elide any other type arguments you wish. When you do so, the missing arguments will be _inferred_ (and will not cause an error to be issued even if they do not have a default)

```ts
const instance4 = new Foo<U = string>(0, "");
const result4 = foo<U = string>(0, "");
const tagged4 = tag<U = string>`tags ${12} ${""}`;
const jsx4 = <Component<U = string> x={12} y="" cb={props => void (props.x.toFixed() + props.y.toUpperCase())} />;
```

(This is not valid for typespace references such as type references - those still have strict arity checks as there is no inference source)

Fixes #22631
Fixes #20122
Fixes #10571

🚲 🏠: Should named type argument assignments use a `=` (as they do in this PR now), or a `:` (as #22631 proposed)?

### CODE REVIEW

**Comment by kpdonn** [comment]
**First question**

> (This is not valid for typespace references such as type references - those still have strict arity checks as there is no inference source)

I assume what you mean here is that something like
```ts
type Example<T, U> = { t: T, u: U }
type NotAllowed = Example<U = string> // not allowed because no where to infer T from
```

Would that be allowed though if the missing type arguments already have defaults specified? Example:
```ts
type ExampleDefaults<T = never, U = never>
type IsThisAllowed = ExampleDefaults<U = string> // allowed?
```

**Second question**
> When you do so, the missing arguments will be inferred (and will not cause an error to be issued even if they do not have a default)

Does that mean the skipped type arguments will still be inferred even if they have a default? Example:
```ts
declare function test<A = any, B = any>(arg: { a?: A, b?: B }): {a: A, b: B}

const r1 = test<string>({ b: "foo" })
const r2 = test<A = string>({ b: "foo" })
```
What are the types of `r1` and `r2` in those cases? Today the type for `r1` is unfortunately `{ a: string, b: any}` because explicitly passing one type argument causes type inference to be skipped for all of them.

**Comment by weswigham** [comment]
> Would that be allowed though if the missing type arguments already have defaults specified? 

Yes, that should be fine. Though you've reminded me of a class of error I probably need to add a test and error for:

```ts
type Two<A, B = number> = [A, B]
type Bug = Two<B = string> // should error, A was not provided
```

> Does that mean the skipped type arguments will still be inferred even if they have a default?

Yes,  and if inference fails (ie, there are no inference sites) it still falls back to the default (which is always the case when inferring normal type parameters).

As for your example, in the first call your result is `{ a: string, b: any }`, as no inference currently takes place unless there's a named type argument (though we may change that). In the second, the result should be `{ a: any, b: string}`, as there's no inference drawn for A, so it falls back to its default.

**Comment by kpdonn** [comment]
@weswigham 
Some tests I think might be useful for potentially weird cases I didn't see covered yet:
```ts
declare function testNamingOtherParameters<A = any, B = any>(arg: { a?: A, b?: B }): { a: A, b: B }
const assumingNotAllowed = testNamingOtherParameters<B = A>({ a: "test" })
// I assume error here because A wouldn't be in scope on the right hand side

declare function stillDefaultsIfNoInference<X, A = string, B= number, C=boolean>(arg: { a?: A, b?: B, c?: C, x?: X}): { a: A, b: B, c: C, x: X }
const result1 = stillDefaultsIfNoInference<C = object> ({ b: "test" })
// expect result1 type is {a: string, b: string, c: object, x: {}}

declare function testConstraints<A extends string, B extends A>(arg?: { a?: A[], b?: B[] }): { a: A[], b: B[] }
const expectAllowed1 = testConstraints < B = "x" > ({ a: ["x", "y"] })
const expectAllowed2 = testConstraints < A = "x" | "y" > ({ b: ["x"] })
const expectError1 = testConstraints < B = "z" > ({ a: ["x", "y"] }) // error "z" not in "x" | "y"
const expectError2 = testConstraints < A = "x" | "y" > ({ b: ["x", "y", "z"] }) // error "z" not in "x" | "y"

declare function complexConstraints<A extends string, B extends A, C extends B>(arg: { a?: A[], b?: B[], c?: C[] }): { a: A[], b: B[], c: C[] }
const expectAllowed3 = complexConstraints < A = "x" | "y" | "z" > ({ a: ["x"], c: ["x", "y"] })

const expectError3 = complexConstraints<A = "x" | "y" | "z", C = "x" | "y">({b: ["x"]}) 
// error because B inferred to be "x" so C can't be "x" | "y"

const expectError4 = complexConstraints<A = "x">({c: ["y"]}) 
```

**Comment by weswigham** [comment]
> But based on what you said before that it sounds like the type would be

Oh, yeah, sorry - I read it wrong XD

**Comment by weswigham** [comment]
@kpdonn Thanks for your test cases - I did change up my implementation a bit to handle some of them, however for a few your assumption of an error was incorrect!

For instance, in:
```ts
const expectError3 = complexConstraints<A = "x" | "y" | "z", C = "x" | "y">({b: ["x"]}) 
```
Inference does fail - we're not allowed to infer `"x"` for B when `C` is supposed to extend it. So then `B` is replaced with its constraint, `A`, which is set to `"x" | "y" | "z"`, and then the call resolves just fine (as a `"x"[]` is assignable to a `("x" | "y" | "z")[]`)

```ts
const expectError1 = testConstraints < B = "z" > ({ a: ["x", "y"] }) // error "z" not in "x" | "y"
```
is similar - inference does fail because of the inference not validating against the constraint, but then we use the constraint as the inference result and the call resolve just fine, as an `("x" | "y")[]` is assignable to a `string[]`.

**Comment by kpdonn** [comment]
👍 Cool that makes sense. I thought of those cases specifically because I figured they'd be really hard to infer so I mainly just curious to see what would happen with them.

**Comment by ohjames** [comment]
With this it's great that you can infer `U` in `<T, U>(a: U)` by using `blah<T = string>(1)`. Seems a bit wordy though given `blah<string>(4)` is an error right now. Would much prefer the version that gives you an error right now to be equivalent to the "named type" version.

**Comment by ohjames** [comment]
It just feels arbitrary that using a named type argument turns on inference of unspecified types when using a positional type argument does not.

Implementing this without also having https://github.com/Microsoft/TypeScript/issues/20122 is gonna introduce semantics that I'm pretty sure will confuse many people.

**Comment by alfaproject** [comment]
@ohjames according to the original post that issue is covered in this PR?

**Comment by ohjames** [comment]
@alfaproject As far as I can tell this only turns on inference when a named positional parameter is used, not when a generic function is called with a partial type parameter list and the unspecified parameters can be inferred. That case will still error even with this PR unless I'm missing something?

**Comment by tbrisbane** [comment]
The closest thing to named function arguments at runtime is having a function take a destructured object as an argument. E.g.
```ts
function foobar({ foo, bar }: { foo: number, bar: number }): [typeof foo, typeof bar] {
  return [foo, bar];
}

foobar({ foo: 8, bar: 16 });
// => [8, 16]
```

In this PR however, you guys opted to go for a `TypeParamName = Type` approach to the syntax. E.g.
```ts
function foobar<T, U>({ foo, bar }: { foo: T, bar: U }): [typeof foo, typeof bar] {
  return [foo, bar];
}

// T presumably inferred as number
foobar<U = number>({ foo: 8, bar: 16 });
// => [8, 16]
```

Have you considered matching the object destructuring syntax at the type-level in order to provide a consistent syntactic experience across both compiletime and runtime? E.g. Something like this:
```ts
function foobar<{ foo: T, bar: U }>({ foo, bar }: { foo: T, bar: U }): [typeof foo, typeof bar] {
  return [foo, bar];
}

// We might want to explicitly say to infer here to make this less "magical"
foobar<{ foo: infer, bar: number }>({ foo: 8, bar: 16 });
// => [8, 16]
```
And another example making the type parameters optional:
```ts
function foobar<{ foo: T = never, bar: U = never }>({ foo, bar }: { foo: T, bar: U }): [typeof foo, typeof bar] {
  return [foo, bar];
}

// typeof bar is never
foobar<{ foo: number }>({ foo: 8 });
// => [8, undefined]

// typeof foo is never
foobar<{ bar: number }>({ bar: 16 });
// => [undefined, 16]

// typeof foo and typeof bar are both never
foobar<{}>({});
// => [undefined, undefined]

// typeof foo and typeof bar are both inferred as number
foobar({ foo: 8, bar: 16 });
// => [8, 16]
```

Are there technical reasons you can't do this kind of type-level object destructuring in the general case?

**Comment by Airblader** [comment]
This change is significant for library maintainers. It'll mean that the chosen name for a generic argument suddenly and implicitly becomes part of the public API. 

Has it been considered to make it explicit to allow referring to type arguments like this?

**Comment by ThomasMichon** [comment]
It occurred to me that several of the scenarios involving named optional generic parameters can be made possible through the use of conditional types:

```ts
interface ITest<A, B> {
    foo: A;
    bar: B;
}

declare function createTest<T = {}>():
    ITest<T extends { a: infer A; } ? A : string, T extends { b: infer B; } ? B : number>;

const test = createTest(); // ITest<string, number>;

const testA = createTest<{ a: boolean; }>(); // ITest<boolean, number>;

const testB = createTest<{ b: boolean; }>(); // ITest<string, boolean>;
```

In the `createTest` function, the specification of the single generic `T` can be used to assign the underlying generics of `ITest` by name, but in a way controlled explicitly by the author of `createTest` with no coupling to `ITest`.

I realize this isn't as complete or as convenient as something which lets you specify the generics of existing interfaces, but it might suggest a direction to take the syntax for this proposal.

**Comment by rubenlg** [comment]
@tbrisbane your proposal has a very specific meaning already in today's Typescript, and now you are adding an extra meaning to the same syntax. For example, look at these two declarations of `foobar`, and how the exact same call means completely different things depending on how the function has been declared:

```typescript
function foobar<T>(...) {...}
function foobar<{ foo: T, bar: U }>(...) {...}

foobar<{ foo: infer, bar: number }>(...);
```
The first problem is that it makes the tsc parser more complex, as the exact same calling code might have two different interpretations depending on how `foobar` is declared. The parser can't construct an AST for the calling code until it has seen the AST for the declaration.

However, my main worry is not the compiler, it's humans, who will have the exact same problem. They will see the call, and not have a clue what that means unless they go and read the declaration.

I prefer the OP proposal, where the two cases are distinct both in how they are declared and how they are used, and there is no room for confusion for the compiler or humans.

**Comment by tbrisbane** [comment]
@rubenlg I'm not sure I follow?

Using your example:

- When the first `foobar` is called, a type variable `T` is instantiated as the provided type in its entirety, i.e. `T = { foo: infer, bar: number }`
- When the second `foobar` is called, type variables `T` and `U` are instantiated as the types of the `foo` and `bar` fields _within_ the provided type, respectively, i.e. `T = infer` and `U = number`

This is no different to the way ES6 value destructuring already works. Take the following for instance:

```js
function barfoo(obj: { foo: string, bar: number }): [typeof obj['foo'], typeof obj['bar']] {
  return [obj.foo, obj.bar];
}
function barfoo({ foo: fooVal, bar: barVal }: { foo: string, bar: number }): [typeof fooVal, typeof barVal] {
  return [fooVal, barVal];
}

barfoo({ foo: 'abc', bar: 42 });
```

That's a value-level equivalent of your example. What problem does your example exhibit that this example does not?

**Comment by tbrisbane** [comment]
Unrelated to my previous comments...

It seems like it may be a better idea to split this work into two components:

1. The ability to leverage inference when manually providing type arguments
    - Examples:
        - `tuple<number, , string>(42, someObject, 'foo')`
        - `tuple<number, ?, string>(42, someObject, 'foo')`
        - `tuple<number, infer, string>(42, someObject, 'foo')`
2. The ability to provide type arguments by name, including leveraging inference
    - Examples:
        - `tuple<T1 = number, T3 = string>(42, someObject, 'foo')`
        - `tuple<T1 = number, T2 = ?, T3 = string>(42, someObject, 'foo')`
        - `tuple<T1 = number, T2 = infer, T3 = string>(42, someObject, 'foo')`

Doing this would help establish consistency around leveraging inference when providing type arguments, regardless of whether you're providing them in order or by some other means (i.e. by name).

Thoughts @weswigham? Or am I completely misunderstanding the kind of feature separation you're going for here? :P

**Comment by rubenlg** [comment]
@tbrisbane Good point. I see where you are coming from. Let me elaborate on my comment a bit more clarifying the differences that I see with destructuring.

In the case of ES6 value destructuring, the meaning at the call site doesn't change. You are always passing an object to the function. Whether the function uses destructuring for convenience or not, is an implementation detail that the caller of the function doesn't need to be aware of. The meaning at the call site is always the same, you call a function with an object.

Neither the Typescript compiler nor humans need to be aware of whether destructuring is used or not to process that function call, as the function _contract_ is the same. TS models the types of the two functions slightly differently (because TS insists on keeping argument names on function types) but that doesn't matter, they are fully assignable to one another.

Your proposal doesn't have that property. The function call has different meanings depending on how the function is declared. The types of the two functions in my example above are very different, both for humans and for the compiler. One takes one generic, with any shape (even a number or string). The other one takes two named generics, but there is no structure to them (no object holding them together).

The call site in my example shows how both functions can be called with the same exact code, but these have very different interpretations:
1. When calling the first function, it's defining a type with structure (it has to be an object containing two fields with specific names and types).
2. When calling the second function it's providing specific types for two named generics, without any structure (no object type is defined containing fields).

That's what makes me uncomfortable, the fact that when reading the function call I (and the compiler) don't know if an object type is being declared, or just two generic types. Argument destructuring doesn't have that problem: from the caller perspective it's always an object.

**Comment by tbrisbane** [comment]
@rubenlg I see what you mean. Thanks for elaborating! :)

**Comment by AnyhowStep** [comment]
I have a lot of generic type parameters in my classes for a project of mine, and I do something similar to what @tbrisbane mentioned.

I'd have, maybe,

```ts
type Data = {
    field0 : type0,
    field1 : type1,
    /*snip*/
};
class ImmutableFoo<DataT extends DataT> {
    //creates new ImmutableFoo<> with data changed
    bar () : ImmutableFoo<{
        [key in keyof DataT] : (
            key extends "field5" ?
            SomeNewType :
            key extends "field22" ?
            SomeNewType2 :
            //Keep old type
            DataT[key]
        )
    }>;
}
```

I'm not sure if this will make writing these kinds of classes easier but I'm hopeful.

**Comment by kourge** [comment]
@Airblader brings up a very good point. Given this revelation, I am motivated to update several libraries of mine to export more descriptive type variable names, lest a user has to deal with nondescript names like `T`, `U`, `V`.

To reiterate: prior to this, changing the type variable names would not require any call site updates, whereas after this, it can cause compilation errors if named type arguments are used.

**Comment by isiahmeadows** [comment]
Just a thought: I've been doing this for a while to great success:

```ts
type SomeType<T extends {Foo: any, Bar: any, ...}> =
    // do things with T["Foo"], T["Bar"], etc.

// later
function foo(): SomeType<{
    Foo: Foo;
    Bar: Bar;
    // ...
}>;
```

I'm not convinced the named type parameter idea is really useful, especially since you can still pass an object parameter without actually creating the object itself. If anything, being able to "destructure" a type parameter would probably be the most useful.

**Comment by treybrisbane** [comment]
@isiahmeadows Yeah those are pretty much my thoughts as well

**Comment by simonbuchan** [comment]
@weswigham Looks like [the roadmap](https://github.com/Microsoft/TypeScript/wiki/Roadmap) should be updated.

**Comment by lewisl9029** [comment]
I tried the approach that @isiahmeadows posted earlier, and it works quite well, with one caveat: I can't seem to figure out how to apply different default generic arguments for each "named" argument.

@isiahmeadows @treybrisbane Any ideas on how default generic args could be implement in this pattern? 

I actually tried to create a helper type that could be applied at the usage sites to supply a default value if a generic type is not supplied, but [it doesn't seem to work as I'd expect it to](https://www.typescriptlang.org/play/index.html#src=export%20type%20WithDefault%3CT%20extends%20%7B%20Type%3A%20any%3B%20Default%3A%20any%20%7D%3E%20%3D%20T%5B%22Type%22%5D%20extends%20void%0A%20%20%3F%20T%5B%22Default%22%5D%0A%20%20%3A%20T%5B%22Type%22%5D%3B%0A%0Aconst%20testFn%20%3D%20%3CT%20extends%20%7B%20DefaultedProp%3F%3A%20any%20%7D%3E(%7B%0A%20%20defaultedProp%0A%7D%3A%20%7B%0A%20%20defaultedProp%3F%3A%20WithDefault%3C%7B%20Type%3A%20T%5B%22DefaultedProp%22%5D%2C%20Default%3A%20string%20%7D%3E%0A%7D)%20%3D%3E%20defaultedProp%0A%0A%2F%2F%20this%20infers%20the%20resulting%20type%20correctly%20as%20number%0Aconst%20testNumber%20%3D%20testFn%3C%7B%20DefaultedProp%3A%20number%20%7D%3E(%7B%20defaultedProp%3A%201234%20%7D)%0A%0A%2F%2F%20this%20doesn't%20seem%20to%20infer%20the%20default%20type%20of%20string%20correctly%0A%2F%2F%20it%20infers%20to%20unknown%0Aconst%20testDefault%20%3D%20testFn%3C%7B%20%7D%3E(%7BdefaultedProp%3A%20'test'%7D)%0A%0Atype%20Test%20%3D%20WithDefault%3C%7B%20Type%3A%20void%2C%20Default%3A%20string%7D%3E).

**Comment by isiahmeadows** [comment]
@lewisl9029 You can't really without defining a type alias injecting the defaults. Something like this might work if you want a general, cookie-cutter solution:

```ts
type Defaults<T, D> = T & Pick<D, Exclude<keyof D, keyof T>>

type Foo<T> = FooImpl<Defaults<T, { /* Your defaults */ }>>
type FooImpl<T> = ... // Your actual type implementation
```

I've never used it nor had a significant need - most of my needs were either all optional or default `undefined`, where `T["key"]` is sufficient.

**Comment by AnyhowStep** [comment]
@lewisl9029 I generally work with the builder pattern when working with generics that have very complex structures.

You get a builder instance and it has sane defaults for most/all of the fields.
Then, you call methods that create a completely new instance of the builder that mutate the type.

At the end of it, you should have an instance of a type that has sane defaults and is highly configurable.
